server/src/main/java/org/apache/kafka/security/authorizer/AclEntry.java

```diff
@@ -0,0 +1,215 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.kafka.common.acl.AclOperation.ALTER;
+import static org.apache.kafka.common.acl.AclOperation.ALTER_CONFIGS;
+import static org.apache.kafka.common.acl.AclOperation.CLUSTER_ACTION;
+import static org.apache.kafka.common.acl.AclOperation.CREATE;
+import static org.apache.kafka.common.acl.AclOperation.CREATE_TOKENS;
+import static org.apache.kafka.common.acl.AclOperation.DELETE;
+import static org.apache.kafka.common.acl.AclOperation.DESCRIBE;
+import static org.apache.kafka.common.acl.AclOperation.DESCRIBE_CONFIGS;
+import static org.apache.kafka.common.acl.AclOperation.DESCRIBE_TOKENS;
+import static org.apache.kafka.common.acl.AclOperation.IDEMPOTENT_WRITE;
+import static org.apache.kafka.common.acl.AclOperation.READ;
+import static org.apache.kafka.common.acl.AclOperation.WRITE;
+
+public class AclEntry extends AccessControlEntry {
+    private static final DecodeJson.DecodeInteger INT = new DecodeJson.DecodeInteger();
+    private static final DecodeJson.DecodeString STRING = new DecodeJson.DecodeString();
+
+    public static final KafkaPrincipal WILDCARD_PRINCIPAL = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "*");
+    public static final String WILDCARD_PRINCIPAL_STRING = WILDCARD_PRINCIPAL.toString();
+    public static final String WILDCARD_HOST = "*";
+    public static final String WILDCARD_RESOURCE = ResourcePattern.WILDCARD_RESOURCE;
+    public static final String RESOURCE_SEPARATOR = ":";
+    public static final Set<ResourceType> RESOURCE_TYPES = Arrays.stream(ResourceType.values())
+        .filter(t -> !(t == ResourceType.UNKNOWN || t == ResourceType.ANY))
+        .collect(Collectors.toSet());
+    public static final Set<AclOperation> ACL_OPERATIONS = Arrays.stream(AclOperation.values())
+        .filter(t -> !(t == AclOperation.UNKNOWN || t == AclOperation.ANY))
+        .collect(Collectors.toSet());
+
+    private static final String PRINCIPAL_KEY = "principal";
+    private static final String PERMISSION_TYPE_KEY = "permissionType";
+    private static final String OPERATION_KEY = "operation";
+    private static final String HOSTS_KEY = "host";
+    public static final String VERSION_KEY = "version";
+    public static final int CURRENT_VERSION = 1;
+    private static final String ACLS_KEY = "acls";
+
+    public final AccessControlEntry ace;
+    public final KafkaPrincipal kafkaPrincipal;
+
+    public AclEntry(AccessControlEntry ace) {
+        super(ace.principal(), ace.host(), ace.operation(), ace.permissionType());
+        this.ace = ace;
+
+        kafkaPrincipal = ace.principal() == null
+            ? null
+            : SecurityUtils.parseKafkaPrincipal(ace.principal());
+    }
+
+    public static AclEntry apply(KafkaPrincipal principal,
```

Comment:
Do we really need this helper? There is only one caller in production code. Also, testing code can create `AccessControlEntry` to instantiate `AclEntry`

Reply:
Removed